Retry fails with injected parameters

### TestNG Version
6.14.2

### Expected behavior
Output when testing against version 6.11.  Notice Test5 is retried three times.
```
SingleSuite > RetryAnalyzer > test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts STANDARD_OUT
    someTest1 Called : Test9
    someTest1 Called : Test8
    someTest1 Called : Test3
    someTest1 Called : Test7
    someTest1 Called : Test6
    someTest1 Called : Test0
    someTest1 Called : Test2
    someTest1 Called : Test5
    someTest1 Called : Test5
    someTest1 Called : Test5
    someTest1 Called : Test5
    someTest1 Called : Test4
    someTest1 Called : Test1
Executed test ==> test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts() with result: SUCCESS
```

### Actual behavior
Output when using 6.14.2.  Notice Test5 is not retried.
```
SingleSuite > RetryAnalyzer > test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts STANDARD_OUT
    someTest1 Called : Test1
    someTest1 Called : Test0
    someTest1 Called : Test5
    someTest1 Called : Test8
    someTest1 Called : Test4
    someTest1 Called : Test3
    someTest1 Called : Test2
    someTest1 Called : Test7
    someTest1 Called : Test9
    someTest1 Called : Test6

SingleSuite > RetryAnalyzer > test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts FAILED
    java.lang.AssertionError at RetryAnalyzerTest.java:29

3 tests completed, 1 failed
```

### Is the issue reproductible on runner?

- [ X] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
Issue can be reproduced using `test.retryAnalyzer.RetryAnalyzerTest` after adding parameter injection to someTest1
```java
    @Test(retryAnalyzer = MyRetry.class)
    public void someTest1(Method method) {
        System.out.println(method.getName() + " Called : " + this.name);
        if (name.contains("5")) {
            m_count++;
            fail();
        }
    }
```

`someTest1` is found in [FactoryTest.java#L23](https://github.com/cbeust/testng/blob/6.14.2/src/test/java/test/retryAnalyzer/FactoryTest.java#L23)
Here's the diff I used to reproduce the issue: https://github.com/heschulz/testng/commit/a4417ded6a0d152965f83f36921325298b96a88f

Here's the full output running RetryAnalyzerTest with this commit...
```
ESCHULZ-M-40ZB:testng eschulz$ ./gradlew test
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :test 
objc[79896]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.

Gradle Test Executor 1 STANDARD_OUT
    ...
    ... TestNG 6.15.0-SNAPSHOT by Cédric Beust (cedric@beust.com)
    ...


SingleSuite > RetryAnalyzer > test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts STANDARD_OUT
    someTest1 Called : Test4
    someTest1 Called : Test2
    someTest1 Called : Test1
    someTest1 Called : Test3
    someTest1 Called : Test5
    someTest1 Called : Test0
    someTest1 Called : Test7
    someTest1 Called : Test8
    someTest1 Called : Test9
    someTest1 Called : Test6

SingleSuite > RetryAnalyzer > test.retryAnalyzer.RetryAnalyzerTest.testInvocationCounts FAILED
    java.lang.AssertionError at RetryAnalyzerTest.java:29

3 tests completed, 1 failed


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///Users/eschulz/ws/testng/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 9s
6 actionable tasks: 1 executed, 5 up-to-date
```
The test will pass with the same diff applied to testng 6.11.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,7 @@
 Current
+Fixed: GITHUB-1706: Retrying of methods fail when test method involves native injection (Krishnan Mahadevan)
+
+6.14.3
 Fixed: GITHUB-1077: TestNG cannot handle load (Aheiss)
 Fixed: GITHUB-1081: group-by-instances with test dependencies causes instantiation of tests to exponentially slow (Aheiss)
 Fixed: GITHUB-1700: Test ignored if @BeforeMethod in base class fails for another test class (Krishnan Mahadevan)

--- a/kobalt/src/Build.kt
+++ b/kobalt/src/Build.kt
@@ -15,7 +15,7 @@ import org.apache.maven.model.Model
 import org.apache.maven.model.Scm
 import java.io.File
 
-val VERSION = "6.14.3"
+val VERSION = "6.15.0-SNAPSHOT"
 
 val p = project {
     name = "testng"

--- a/src/main/java/org/testng/internal/Invoker.java
+++ b/src/main/java/org/testng/internal/Invoker.java
@@ -814,6 +814,7 @@ public class Invoker implements IInvoker {
 
   int retryFailed(Object instance,
                            ITestNGMethod tm,
+                           Object[] paramValues,
                            XmlSuite suite,
                            ITestClass testClass,
                            ITestNGMethod[] beforeMethods,
@@ -835,6 +836,9 @@ public class Invoker implements IInvoker {
 
       ParameterBag bag = handler.createParameters(tm, parameters, allParameters, testContext);
       Object[] parameterValues = Parameters.getParametersFromIndex(bag.parameterHolder.parameters, parametersIndex);
+      if (bag.parameterHolder.origin == ParameterHolder.ParameterOrigin.NATIVE) {
+        parameterValues = paramValues;
+      }
 
       result.add(invokeMethod(instance, tm, parameterValues, parametersIndex, suite,
           allParameters, testClass, beforeMethods, afterMethods, groupMethods, failure));
@@ -1008,7 +1012,7 @@ public class Invoker implements IInvoker {
                 List<ITestResult> retryResults = Lists.newArrayList();
 
                 failure.count = retryFailed(
-                        instance, testMethod, suite, testClass, beforeMethods,
+                        instance, testMethod, parameterValues, suite, testClass, beforeMethods,
                         afterMethods, groupMethods, retryResults,
                         failure.count,
                         testContext, parameters, parametersIndex);

--- a/src/main/java/org/testng/internal/ParameterHolder.java
+++ b/src/main/java/org/testng/internal/ParameterHolder.java
@@ -16,7 +16,8 @@ public class ParameterHolder {
    */
   public enum ParameterOrigin {
     ORIGIN_DATA_PROVIDER, // A data provider
-    ORIGIN_XML // TestNG XML suite
+    ORIGIN_XML, // TestNG XML suite
+    NATIVE //Native injection is involved.
   }
 
   public IDataProviderMethod dataProviderHolder;

--- a/src/main/java/org/testng/internal/Parameters.java
+++ b/src/main/java/org/testng/internal/Parameters.java
@@ -690,6 +690,7 @@ public class Parameters {
     final IDataProviderMethod dataProviderMethod =
             findDataProvider(instance, testMethod.getTestClass(),
                     testMethod.getConstructorOrMethod(), annotationFinder, methodParams.context);
+    ParameterOrigin origin;
 
     if (null != dataProviderMethod) {
       int parameterCount = testMethod.getConstructorOrMethod().getParameterTypes().length;
@@ -758,8 +759,12 @@ public class Parameters {
       });
 
       return new ParameterHolder(filteredParameters, ParameterOrigin.ORIGIN_DATA_PROVIDER, dataProviderMethod);
+    } else if (methodParams.xmlParameters.isEmpty() ) {
+      origin = ParameterOrigin.NATIVE;
     }
     else {
+      origin = ParameterOrigin.ORIGIN_XML;
+    }
       //
       // Normal case: we have only one set of parameters coming from testng.xml
       //
@@ -776,8 +781,7 @@ public class Parameters {
       // Turn it into an Iterable
       Iterator<Object[]> parameters = new ArrayIterator(allParameterValuesArray);
 
-      return new ParameterHolder(parameters, ParameterOrigin.ORIGIN_XML, null);
-    }
+      return new ParameterHolder(parameters, origin, null);
   }
 
   /**

--- a/src/main/java/org/testng/internal/TestMethodWithDataProviderMethodWorker.java
+++ b/src/main/java/org/testng/internal/TestMethodWithDataProviderMethodWorker.java
@@ -90,7 +90,7 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
 
           m_failureCount =
              m_invoker.retryFailed(
-                 instance, m_testMethod, m_xmlSuite, m_testClass, m_beforeMethods,
+                 instance, m_testMethod, m_parameterValues, m_xmlSuite, m_testClass, m_beforeMethods,
                  m_afterMethods, m_groupMethods, retryResults,
                  m_failureCount,
                  m_testContext, m_parameters, m_parameterIndex);

--- a/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
+++ b/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
@@ -2,19 +2,29 @@ package test.retryAnalyzer;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 
 import org.testng.ITestNGListener;
 import org.testng.ITestResult;
 import org.testng.TestListenerAdapter;
 import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
+import org.testng.collections.Maps;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+import test.InvokedMethodNameListener;
 import test.SimpleBaseTest;
 import test.retryAnalyzer.github1519.MyListener;
 import test.retryAnalyzer.github1519.TestClassSample;
 import test.retryAnalyzer.github1600.Github1600Listener;
 import test.retryAnalyzer.github1600.Github1600TestSample;
+import test.retryAnalyzer.github1706.DataDrivenSample;
+import test.retryAnalyzer.github1706.NativeInjectionSample;
+import test.retryAnalyzer.github1706.ParameterInjectionSample;
 
 public class RetryAnalyzerTest extends SimpleBaseTest {
     @Test
@@ -39,7 +49,7 @@ public class RetryAnalyzerTest extends SimpleBaseTest {
     @Test
     public void testIfRetryIsInvokedBeforeListener() {
         TestNG tng = create(TestClassSample.class);
-        tng.addListener((ITestNGListener)new MyListener());
+        tng.addListener((ITestNGListener) new MyListener());
         tng.run();
         assertThat(TestClassSample.messages).containsExactly("afterInvocation", "retry", "afterInvocation");
     }
@@ -55,4 +65,35 @@ public class RetryAnalyzerTest extends SimpleBaseTest {
         assertThat(tla.getFailedTests()).hasSize(1);
         assertThat(tla.getSkippedTests()).hasSize(1);
     }
+
+    @Test(description = "GITHUB-1706", dataProvider = "1706")
+    public void testIfRetryIsInvokedWhenTestMethodHas(Class<?> clazz, int size, Map<String, String> parameters) {
+        XmlSuite xmlsuite = createXmlSuite("suite");
+        XmlTest xmlTest = createXmlTest(xmlsuite, "test", clazz);
+        if (!parameters.isEmpty()) {
+            xmlTest.setParameters(parameters);
+        }
+        TestNG tng = create();
+        tng.setXmlSuites(Collections.singletonList(xmlsuite));
+        InvokedMethodNameListener listener = new InvokedMethodNameListener();
+        tng.addListener((ITestNGListener) listener);
+        tng.run();
+        assertThat(listener.getSkippedMethodNames().size()).isEqualTo(size);
+    }
+
+    @DataProvider(name = "1706")
+    public Object[][] getData() {
+        return new Object[][]{
+                {NativeInjectionSample.class, 2, Maps.newHashMap()},
+                {DataDrivenSample.class, 4, Maps.newHashMap()},
+                {ParameterInjectionSample.class, 2, constructParameterMap()}
+        };
+    }
+
+    private static Map<String, String> constructParameterMap() {
+        Map<String, String> map = Maps.newHashMap();
+        map.put("counter", "3");
+        return map;
+    }
+
 }

--- a/src/test/java/test/retryAnalyzer/github1706/DataDrivenSample.java
+++ b/src/test/java/test/retryAnalyzer/github1706/DataDrivenSample.java
@@ -1,0 +1,29 @@
+package test.retryAnalyzer.github1706;
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DataDrivenSample {
+    private AtomicInteger counter = new AtomicInteger(0);
+
+    @Test(retryAnalyzer = LocalRetry.class, dataProvider = "getdata")
+    public void testMethod(int i) {
+        Assert.assertTrue( i > 0);
+        if (counter.incrementAndGet() != 3) {
+            Assert.fail();
+        }
+        counter = new AtomicInteger(0);
+    }
+
+    @DataProvider(name = "getdata")
+    public Object[][] getData() {
+        return new Object[][]{
+                {1},
+                {2}
+        };
+    }
+
+}

--- a/src/test/java/test/retryAnalyzer/github1706/LocalRetry.java
+++ b/src/test/java/test/retryAnalyzer/github1706/LocalRetry.java
@@ -1,0 +1,11 @@
+package test.retryAnalyzer.github1706;
+
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+public class LocalRetry implements IRetryAnalyzer {
+    @Override
+    public boolean retry(ITestResult result) {
+        return true;
+    }
+}

--- a/src/test/java/test/retryAnalyzer/github1706/NativeInjectionSample.java
+++ b/src/test/java/test/retryAnalyzer/github1706/NativeInjectionSample.java
@@ -1,0 +1,19 @@
+package test.retryAnalyzer.github1706;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.lang.reflect.Method;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class NativeInjectionSample {
+    private AtomicInteger counter = new AtomicInteger(0);
+
+    @Test(retryAnalyzer = LocalRetry.class)
+    public void testMethod(Method method) {
+        Assert.assertNotNull(method);
+        if (counter.incrementAndGet() != 3) {
+            Assert.fail();
+        }
+    }
+}

--- a/src/test/java/test/retryAnalyzer/github1706/ParameterInjectionSample.java
+++ b/src/test/java/test/retryAnalyzer/github1706/ParameterInjectionSample.java
@@ -1,0 +1,20 @@
+package test.retryAnalyzer.github1706;
+
+import org.testng.Assert;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ParameterInjectionSample {
+    private AtomicInteger counter = new AtomicInteger(0);
+
+    @Test(retryAnalyzer = LocalRetry.class)
+    @Parameters({"counter"})
+    public void testMethod(int paramCounter) {
+        Assert.assertTrue(paramCounter > 0);
+        if (counter.incrementAndGet() != 3) {
+            Assert.fail();
+        }
+    }
+}

